### PR TITLE
Fedora 31

### PR DESCRIPTION
**Pull request by Gundersanne**
* [x] image-refresh fedora-31
- [x] rebuild image again to get grubby installed
- [x] fix custom cockpit keytab: PR #12732

Issues:
- atomic  package isn't available on f31
- docker doesn't support cgroupsv2: https://github.com/cockpit-project/cockpit/issues/12670

### CODE REVIEW

File: bots/image-refresh [hunk 1]
```diff
@@ -57,6 +57,8 @@ TRIGGERS = {
         "fedora-30/firefox@weldr/cockpit-composer",
         "fedora-30/edge@weldr/cockpit-composer",
     ],
+    "fedora-31": [
```
Comment (marusak) [inline bots/image-refresh]: This commit should be squashed into the main one. We need to separately declare in `testmap.py` and land that before we can test this PR.

**Comment by marusak** [comment]
> 5. Needs to build image in this PR

I see you added it. Don't forget to add `bot` label when ready so bots really pick it up

**Comment by cockpituous** [comment]
image-refresh in progress on 3-ci-srv-06.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190822-092626/

File: bots/images/scripts/fedora-31.bootstrap [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+# General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with this program; if not, write to the Free Software
+# Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+# 02110-1301 USA.
+
+BASE=$(dirname $0)
+$BASE/virt-install-fedora "$1" x86_64 "https://dl.fedoraproject.org/pub/fedora/linux/development/31/Everything/x86_64/os/"
```
Comment (marusak) [inline bots/images/scripts/fedora-31.bootstrap]: aand one more thing here please :)
Add in comment the URL when it is released. It is much easier to switch to stable release then.
See https://github.com/cockpit-project/cockpit/pull/11432/files#diff-29144c315de7844b0f831b03a2404e1bR21

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190822-092626/

**Comment by cockpituous** [comment]
image-refresh in progress on 1-ci-srv-03.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190822-122626/

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190822-122626/

**Comment by cockpituous** [comment]
image-refresh in progress on centosci-tasks-w4dn3.
Log: https://logs-https-cockpit.apps.ci.centos.org/logs/image-refresh-12619-20190823-074646/

**Comment by cockpituous** [comment]
Task failed: https://logs-https-cockpit.apps.ci.centos.org/logs/image-refresh-12619-20190823-074646/

**Comment by cockpituous** [comment]
image-refresh in progress on 3-cockpit-8.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190823-081515/

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190823-081515/

**Comment by cockpituous** [comment]
image-refresh in progress on 2-ci-srv-01.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-065151/

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-065151/

**Comment by cockpituous** [comment]
image-refresh in progress on 4-ci-srv-05.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-073232/

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-073232/

**Comment by cockpituous** [comment]
image-refresh in progress on 3-ci-srv-04.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-081212/

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-081212/

**Comment by cockpituous** [comment]
image-refresh in progress on 3-cockpit-10.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-112727/

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-112727/

**Comment by cockpituous** [comment]
image-refresh in progress on 4-ci-srv-05.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-120000/

**Comment by cockpituous** [comment]
Task failed: https://209.132.184.41:8493/logs/image-refresh-12619-20190826-120000/

**Comment by cockpituous** [comment]
image-refresh in progress on 4-ci-srv-01.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190827-090101/

**Comment by cockpituous** [comment]
Couldn't configure git config with our API token: https://209.132.184.41:8493/logs/image-refresh-12619-20190827-090101/

**Comment by cockpituous** [comment]
image-refresh in progress on 3-cockpit-8.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190827-125757/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190827-140505

**Comment by cockpituous** [comment]
image-refresh in progress on centosci-tasks-0c7mc.
Log: https://logs-https-cockpit.apps.ci.centos.org/logs/image-refresh-12619-20190830-161212/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190830-162727

**Comment by cockpituous** [comment]
image-refresh in progress on centosci-tasks-n1r4n.
Log: https://logs-https-cockpit.apps.ci.centos.org/logs/image-refresh-12619-20190830-163737/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190830-165353

**Comment by martinpitt** [comment]
@Gundersanne: Wrt. missing "atomic", sounds related to https://bugzilla.redhat.com/show_bug.cgi?id=1746750

**Comment by Gundersanne** [comment]
@martinpitt  Maybe I misunderstood but I thought that was deprecated in favour of silverblue.

**Comment by martinpitt** [comment]
@Gundersanne: That's right. Although there is still a difference between Atomic, the distro product, and "atomic", the package (containing a wrapper around docker, essentially). This will remove some functionality around setting up docker storage (which calls the `atomic` CLI), but it seems this is being deprecated as a whole anyway in favor of docker. So it seems fine to just disable these parts on F31.

**Comment by Gundersanne** [comment]
@martinpitt Ah, but docker is not available right now on fedora-31 because the cgroup2 support. So it doesn't have anything to wrap I guess.

**Comment by martinpitt** [comment]
@Gundersanne: Heh, fun -- so let's just skip all the docker tests on F31 for now? This needs some research whether it's aimed at getting fixed in time for the release (which is only a few weeks out!), if not we should just stop building cockpit-docker entirely on Fedora ≥ 31.

**Comment by Gundersanne** [comment]
@martinpitt afaik it's being tracked here https://github.com/opencontainers/runc/issues/654 so doesn't look like it

**Comment by martinpitt** [comment]
@Gundersanne : Hm, still good question whether docker will switch to `crun` in time, or just stop working.

**Comment by cockpituous** [comment]
image-refresh in progress on 1-ci-srv-02.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190902-184747/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190902-193030

**Comment by cockpituous** [comment]
Failed to confirm the presence of all triggers: https://209.132.184.41:8493/logs/image-refresh-12619-20190902-184747/

**Comment by cockpituous** [comment]
image-refresh in progress on 1-ci-srv-03.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190902-201111/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190902-205757

**Comment by cockpituous** [comment]
Failed to confirm the presence of all triggers: https://209.132.184.41:8493/logs/image-refresh-12619-20190902-201111/

**Comment by cockpituous** [comment]
image-refresh in progress on 4-cockpit-7.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190903-082323/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190903-091818

**Comment by cockpituous** [comment]
Failed to confirm the presence of all triggers: https://209.132.184.41:8493/logs/image-refresh-12619-20190903-082323/

**Comment by cockpituous** [comment]
image-refresh in progress on centosci-tasks-d4sk4.
Log: https://logs-https-cockpit.apps.ci.centos.org/logs/image-refresh-12619-20190903-092929/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190903-094545

**Comment by cockpituous** [comment]
image-refresh in progress on 2-ci-srv-02.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190904-152020/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190904-165151

**Comment by cockpituous** [comment]
image-refresh in progress on 1-ci-srv-05.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190905-144545/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190905-155353

**Comment by martinpitt** [comment]
Notes about the curl check-realms failure: My initial thought was that F31 may have switched from the kernel keyring to the user-land KCM (keyring cache daemon) by default. But /etc/krb5.conf.d/kcm_default_ccache already exists in F30, and so does sssd-kcm.service. On F31 this  service is active, and in `klist` I do see the ticket:
```
# klist
Ticket cache: KCM:0
Default principal: admin@COCKPIT.LAN

Valid starting       Expires              Service principal
06.09.2019 10:02:54  07.09.2019 10:02:53  host/x0.cockpit.lan@COCKPIT.LAN
06.09.2019 10:02:53  07.09.2019 10:02:53  krbtgt/COCKPIT.LAN@COCKPIT.LAN
06.09.2019 10:03:01  07.09.2019 10:02:53  HTTP/x0.cockpit.lan@COCKPIT.LAN
```

But with curl

    curl --negotiate --delegation always -u : -D- http://x0.cockpit.lan:9090/cockpit/login

you get 

    cockpit-session: gssapi auth failed: Request ticket server HTTP/x0.cockpit.lan@COCKPIT.LAN not found in keytab (ticket kvno 1)

But it's cleary there:
```
# klist -k /etc/cockpit/krb5.keytab 
Keytab name: FILE:/etc/cockpit/krb5.keytab
KVNO Principal
---- --------------------------------------------------------------------------
   1 HTTP/x0.cockpit.lan@COCKPIT.LAN
   1 HTTP/x0.cockpit.lan@COCKPIT.LAN
```
This is pretty much exactly how it looks like on Fedora 30.

FTR, this is *not* about the user ticket, that works fine (with KCM and kernel keyring, and you get a ticket either way). This is about kerberos not being able to authenticate the cockpit service.

**Comment by cockpituous** [comment]
image-refresh in progress on 2-cockpit-9.
Log: https://209.132.184.41:8493/logs/image-refresh-12619-20190906-162424/

**Comment by cockpituous** [comment]
image-refresh fedora-31 done: https://github.com/cockpituous/cockpit/commits/image-refresh-fedora-31-20190906-170808

**Comment by martinpitt** [comment]
I sent a fix for custom krb keytab in PR #12732. Adding "blocked" for that. Feel free to rope in the commit in the meantime if you wan to run tests on CI.

**Comment by martinpitt** [comment]
Unblocking. PR #12732 landed, so kerberos (check-realms) should work now.

File: bots/images/scripts/fedora.setup [hunk 3]
```diff
@@ -17,7 +17,9 @@ fi
 # depending on how it was compiled and which plugins are installed,
 # and will just silently not do it.  So we do it here.
 #
-xfs_growfs /
+if [ "$IMAGE" != fedora-31 ]; then
+    xfs_growfs /
+fi
```
Comment (martinpitt) [inline bots/images/scripts/fedora.setup]: This seems awfully specific -- what's the difference here, F31 switched to something else than XFS? Or can this resizing be dropped for other images as well?

File: bots/images/scripts/fedora.setup [hunk 4]
```diff
@@ -60,6 +62,12 @@ tuned \
 virt-install \
 "
 
+# HACK docker not available on f31
+# https://github.com/cockpit-project/cockpit/issues/12670
+if [ "$IMAGE" != fedora-31 ]; then
+    COCKPIT_DEPS="$COCKPIT_DEPS atomic docker"
+fi
```
Comment (martinpitt) [inline bots/images/scripts/fedora.setup]: If that's a temporary hack, fine. If it's planned (or at least likely) that docker just won't be a thing any more in F31, then let's swap this around: Only add docker to fedora-30 (legacy mode).

File: bots/images/scripts/fedora.setup [hunk 5]
```diff
@@ -164,23 +172,25 @@ if ! [ "$version" -eq "$version" ] 2>/dev/null; then version=30; fi
...
 
-# HACK: docker falls over regularly, print its log if it does
-systemctl start docker || journalctl -u docker
-
-# our cockpit/base container is only really a thing on x86_64, just skip it on other arches
-if [ $(uname -m) = x86_64 ]; then
-    docker build -t cockpit/base /var/tmp/cockpit-base
+# HACK docker not available on f31
+# https://github.com/cockpit-project/cockpit/issues/12670
+if [ "$IMAGE" != fedora-31 ]; then
```
Comment (martinpitt) [inline bots/images/scripts/fedora.setup]: Let's perhaps make this `if type docker >/dev/null 2>&1`, as long as it's not yet clear which relelases docker will end up on in the long run. Or something similar to `$HAVE_KUBERNETES`?

File: bots/images/scripts/fedora.setup [hunk 6]
```diff
@@ -164,23 +172,25 @@ if ! [ "$version" -eq "$version" ] 2>/dev/null; then version=30; fi
...
-# docker images that we need for integration testing
-/var/lib/testvm/docker-images.setup
-
 # reduce image size
 dnf clean all
-/var/lib/testvm/zero-disk.setup
+zd_opts=
+if [ "$IMAGE" = fedora-31 ]; then
+    zd_opts="--keep-mock-cache"
+fi
```
Comment (martinpitt) [inline bots/images/scripts/fedora.setup]: This needs an explanation.

File: bots/images/scripts/lib/fedora.install [hunk 7]
```diff
@@ -91,6 +92,9 @@ else
     echo '====== skipping rpmlint check, not installed ====='
 fi
 EOF
+    chmod +x /tmp/rpmlint
+    su builder -c "/usr/bin/mock --offline --copyin /tmp/rpmlint /var/tmp/rpmlint"
+    su builder -c "/usr/bin/mock --offline --shell /var/tmp/rpmlint"
```
Comment (martinpitt) [inline bots/images/scripts/lib/fedora.install]: Can you please split this out into a separate commit (or even PR) with a more detailled explanation?

File: tools/cockpit.spec [hunk 8]
```diff
@@ -206,7 +206,8 @@ echo '%dir %{_datadir}/cockpit/playground' > tests.list
 find %{buildroot}%{_datadir}/cockpit/playground -type f >> tests.list
 
 %ifarch x86_64 %{arm} aarch64 ppc64le i686 s390x
-%if 0%{?fedora}
+# HACK: Docker not available https://github.com/cockpit-project/cockpit/issues/12670
+%if 0%{?fedora} && 0%{?fedora} <= 30
```
Comment (martinpitt) [inline tools/cockpit.spec]: This comes back to the question whether docker is broken temporarily or on its way out of Fedora 31. In the former case we should still build/release the package, as it's more confusing/procedural overhead to remove and re-introduce it after the F31 release.

File: tools/cockpit.spec [hunk 9]
```diff
@@ -634,7 +634,7 @@ bastion hosts, and a basic dashboard.
 %endif
 
 %ifarch x86_64 %{arm} aarch64 ppc64le i686 s390x
-%if 0%{?fedora}
+%if 0%{?fedora} && 0%{?fedora} <= 30
```
Comment (martinpitt) [inline tools/cockpit.spec]: Not sure why this is in the "docker: Skip docker tests on fedora-31" *and* the previous "add fedora-31" commit. (github web UI bug?). Either way, this commit should be squashed into the "add fedora-31 image" commit.

File: bots/naughty/fedora-31/10038-abrtd-crash-fn-2 [hunk 10]
```diff
@@ -0,0 +1,5 @@
+Traceback (most recent call last):
+  File "test/verify/check-journal", line *, in testAbrtDelete
+    b.click(sel)
+*
+Error: timeout
```
Comment (martinpitt) [inline bots/naughty/fedora-31/10038-abrtd-crash-fn-2]: Please squash this "add naughties" commit as well.

File: test/verify/check-networking-firewall [hunk 11]
```diff
@@ -34,6 +34,12 @@ def wait_unit_state(machine, unit, state):
 
     wait(lambda: active_state(unit) == state, delay=0.2)
 
+def get_active_rules(machine):
```
Comment (martinpitt) [inline test/verify/check-networking-firewall]: Most of this commit looks like it could go into a separate PR and be landed first, as it's some generalization? Then the "add fedora-31" main commit can just add the "fedora-31" special cases to the lists.

File: bots/naughty/fedora-31/12730-no-custom-keytab [hunk 12]
```diff
@@ -0,0 +1,3 @@
+File "test/verify/check-realms", line *, in *
+  self.assertIn("HTTP/1.1 200 OK", output)
+AssertionError: 'HTTP/1.1 200 OK' not found in 'HTTP/1.1 401*
```
Comment (martinpitt) [inline bots/naughty/fedora-31/12730-no-custom-keytab]: This should be fixed properly now  in master, and can be dropped.

File: test/verify/check-dashboard [hunk 13]
```diff
@@ -94,7 +94,7 @@ class TestBasicDashboard(MachineCase, DashBoardHelpers):
 
         # Start second browser and check that it is in sync
         b2 = self.new_browser()
-        b2.default_user = "root"
+        b2.default_user = "admin"
```
Comment (martinpitt) [inline test/verify/check-dashboard]: This could also be split out and landed  separately. (If it helps, then you don't need to run the entire set of tests for this PR)

File: bots/images/scripts/fedora.setup (hunk 3 above)
Comment (Gundersanne) [reply bots/images/scripts/fedora.setup]: It's not available (yet) on F31. xfsprogs hasn't been released yet to f31 either.

Comment (martinpitt) [reply bots/images/scripts/fedora.setup]: At least on my slightly older fedora-31 image I see

    /dev/mapper/fedora-root  9.6G  4.8G  4.4G  52% /

Which seems fine. So perhaps that whole thing is obsolete for the other images, too. We can clean these up in a follow-up, so this is fine for now.

**Comment by martinpitt** [comment]
This now conflicts after PR #12735, sorry.

File: tools/cockpit.spec (hunk 8 above)
Comment (Gundersanne) [reply tools/cockpit.spec]: Well it's still unclear, and if it is, it won't be for a while.

File: bots/images/scripts/fedora.setup [hunk 14]
```diff
@@ -27,16 +29,21 @@ if [ $(uname -m) = x86_64 ]; then
     HAVE_KUBERNETES=1
 fi
 
+HAVE_DOCKER=
+if type docker >/dev/null 2>&1; then
+    HAVE_DOCKER=1
+fi
```
Comment (martinpitt) [inline bots/images/scripts/fedora.setup]: Wait, that looks wrong now -- this test is done before the docker package gets installed below. As this is chicken & egg, let's go back to testing `if [ "$1" = fedora-31 ]`.  Sorry for the bogus advice before!

File: bots/images/scripts/fedora.setup [hunk 15]
```diff
@@ -27,16 +29,21 @@ if [ $(uname -m) = x86_64 ]; then
...
 # We install all dependencies of the cockpit packages since we want
 # them to not spontaneously change from one test run to the next when
 # the distribution repository is updated.
 #
 COCKPIT_DEPS="\
-atomic \
 device-mapper-multipath \
-docker \
 etcd \
+glibc-all-langpacks \
```
Comment (martinpitt) [inline bots/images/scripts/fedora.setup]: Why do we need that?

File: bots/images/scripts/fedora.setup [hunk 16]
```diff
@@ -164,23 +175,26 @@ if ! [ "$version" -eq "$version" ] 2>/dev/null; then version=30; fi
...
 
-# HACK: docker falls over regularly, print its log if it does
-systemctl start docker || journalctl -u docker
-
-# our cockpit/base container is only really a thing on x86_64, just skip it on other arches
-if [ $(uname -m) = x86_64 ]; then
-    docker build -t cockpit/base /var/tmp/cockpit-base
+# HACK docker not available on f31
+# https://github.com/cockpit-project/cockpit/issues/12670
+if [ "$HAVE_DOCKER" ]; then
```
Comment (martinpitt) [inline bots/images/scripts/fedora.setup]: not that important here, but let's please use a proper operator: `[ -n "$HAVE_DOCKER" ]`. Also, no need to repeat the comment.

File: bots/images/scripts/fedora.setup (hunk 15 above)
Comment (Gundersanne) [reply bots/images/scripts/fedora.setup]: Otherwise the localization tests fail. These are not installed by default on f31.

Comment (marusak) [reply bots/images/scripts/fedora.setup]: just out of interest, I thought we only tested localization on F30 (but surely we can switch to F31).
